File: bindings/wasm/examples/node/diff_chain.js
// Copyright 2020-2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

const { logExplorerUrl } = require("./explorer_util");
const { Client, Config, Document, Service } = require("../../node/identity_wasm");
const { manipulateIdentity } = require("./manipulate_did");

/**
    This example is a basic introduction to creating a diff message and publishing it to the tangle.
    1. A did document is created and published with one service.
    2. The document is cloned and another service is added.
    3. The difference between the two documents is created and published as a diff message.
    4. The final DID will contain both services.

    @param {{network: string, node: string}} clientConfig
    @param {boolean} log log the events to the output window
**/
async function createDiffChain(clientConfig) {
    // Create a default client configuration from the parent config network.
    const config = Config.fromNetwork(clientConfig.network);

    // Create a client instance to publish messages to the Tangle.
    const client = Client.fromConfig(config);

    // Creates a new identity, that also is updated (See "manipulate_did" example).
    const { doc, key, updatedMessageId } = await manipulateIdentity(clientConfig);

    // clone the Document
    const doc2 = Document.fromJSON(doc.toJSON());

    // Add a second ServiceEndpoint
    let serviceJSON = {
        id: doc.id + "#new-linked-domain",
        type: "LinkedDomains",
        serviceEndpoint: "https://identity.iota.org",
    };
    doc2.insertService(Service.fromJSON(serviceJSON));
    console.log(doc2);

    // Create diff update
    const diff = doc.diff(doc2, updatedMessageId, key);
    console.log(diff);

    // Publish diff to the Tangle
    const diffReceipt = await client.publishDiff(updatedMessageId, diff);
    console.log(diffReceipt);
    logExplorerUrl("Diff Chain Transaction:", clientConfig.network.toString(), diffReceipt.messageId);

    return { doc2, key, diffMessageId: diffReceipt.messageId };
}

exports.createDiffChain = createDiffChain;

File: examples/low-level-api/diff_chain.rs
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! An example that demonstrates how to publish changes to the integration chain to update a
//! DID Document.
//!
//! cargo run --example update_did

use identity::core::json;
use identity::core::FromJson;
use identity::core::Timestamp;
use identity::core::ToJson;
use identity::did::Service;
use identity::did::DID;
use identity::iota::ExplorerUrl;
use identity::iota::Receipt;
use identity::iota_core::DiffMessage;
use identity::iota_core::IotaService;
use identity::prelude::*;

mod create_did;

#[tokio::main]
async fn main() -> Result<()> {
  // Create a client instance to send messages to the Tangle.
  let client: Client = Client::new().await?;

  // Create a signed DID Document and KeyPair (see create_did.rs).
  let (document, keypair, receipt): (IotaDocument, KeyPair, Receipt) = create_did::run().await?;

  // Update the document by adding a new service endpoint
  let updated_document: IotaDocument = {
    let mut doc: IotaDocument = document.clone();

    // Add a Service
    let service: IotaService = Service::from_json_value(json!({
      "id": doc.id().to_url().join("#linked-domain-1")?,
      "type": "LinkedDomains",
      "serviceEndpoint": "https://example.com/"
    }))?;
    assert!(doc.insert_service(service));
    doc.metadata.updated = Timestamp::now_utc();
    doc
  };

  // Generate a signed diff object.
  let diff: DiffMessage = document.diff(
    &updated_document,
    *receipt.message_id(),
    keypair.private(),
    document.default_signing_method()?.id(),
  )?;

  println!("Diff > {}", diff.to_json_pretty().unwrap());

  // Publish the diff object to the Tangle, starting a diff chain.
  let update_receipt: Receipt = client.publish_diff(receipt.message_id(), &diff).await?;

  println!("Diff Update Receipt > {:#?}", update_receipt);

  // Display the web explorer url that shows the published diff message.
  let explorer: &ExplorerUrl = ExplorerUrl::mainnet();
  println!(
    "Diff Update Transaction > {}",
    explorer.message_url(update_receipt.message_id())?
  );
  println!("Explore the DID Document > {}", explorer.resolver_url(document.id())?);

  Ok(())
}
